File: src/main/java/kernitus/plugin/OldCombatMechanics/commands/OCMCommandHandler.java
/*
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/.
 */
package kernitus.plugin.OldCombatMechanics.commands;

import kernitus.plugin.OldCombatMechanics.OCMMain;
import kernitus.plugin.OldCombatMechanics.module.ModuleAttackCooldown;
import kernitus.plugin.OldCombatMechanics.utilities.Config;
import kernitus.plugin.OldCombatMechanics.utilities.Messenger;
import org.bukkit.Bukkit;
import org.bukkit.ChatColor;
import org.bukkit.World;
import org.bukkit.command.Command;
import org.bukkit.command.CommandExecutor;
import org.bukkit.command.CommandSender;
import org.bukkit.configuration.file.FileConfiguration;
import org.bukkit.entity.Player;
import org.bukkit.plugin.PluginDescriptionFile;
import org.jetbrains.annotations.NotNull;

import java.io.File;
import java.util.*;
import java.util.stream.Collectors;

public class OCMCommandHandler implements CommandExecutor {
    private static final String NO_PERMISSION = "&cYou need the permission '%s' to do that!";

    private final OCMMain plugin;
    private final File pluginFile;

    enum Subcommand {reload, toggle, enable, disable}

    public OCMCommandHandler(OCMMain instance, File pluginFile) {
        this.plugin = instance;
        this.pluginFile = pluginFile;
    }

    private void help(OCMMain plugin, CommandSender sender) {
        final PluginDescriptionFile description = plugin.getDescription();

        Messenger.send(sender, ChatColor.DARK_GRAY + Messenger.HORIZONTAL_BAR);
        Messenger.send(sender, "&6&lOldCombatMechanics&e by &ckernitus&e and &cRayzr522&e version &6%s", description.getVersion());

        if (checkPermissions(sender, Subcommand.reload))
            Messenger.send(sender, "&eYou can use &c/ocm reload&e to reload the config file");
        if (checkPermissions(sender, Subcommand.toggle))
            Messenger.send(sender, "&eYou can use &c/ocm toggle [player] [on/off] &e to turn attack cooldown on/off");
        if (checkPermissions(sender, Subcommand.enable) || checkPermissions(sender, Subcommand.disable))
            Messenger.send(sender, "&eYou can use &c/ocm <enable/disable> [world] &e to toggle cooldown for the server or world");

        Messenger.send(sender, ChatColor.DARK_GRAY + Messenger.HORIZONTAL_BAR);


    }

    private void reload(CommandSender sender) {
        Config.reload();
        Messenger.send(sender, "&6&lOldCombatMechanics&e config file reloaded");
    }

    private void toggle(OCMMain plugin, CommandSender sender, String[] args) {
        final FileConfiguration config = plugin.getConfig();

        Player player = null;
        ModuleAttackCooldown.PVPMode mode = null;

        if (args.length >= 2) {
            player = Bukkit.getPlayer(args[1]);

            if (args.length >= 3) {
                if (args[2].equalsIgnoreCase("on")) mode = ModuleAttackCooldown.PVPMode.NEW_PVP;
                else if (args[2].equalsIgnoreCase("off")) mode = ModuleAttackCooldown.PVPMode.OLD_PVP;
            }
        }

        if (player == null && sender instanceof Player) player = (Player) sender;
        if (player == null) {
            final String message = config.getString("disable-attack-cooldown.message-usage",
                    "&4ERROR: &rdisable-attack-cooldown.message-usage string missing");
            Messenger.sendNormalMessage(sender, message);
            return;
        }

        if (mode == null) {
            ModuleAttackCooldown.PVPMode oldMode = ModuleAttackCooldown.PVPMode.getModeForPlayer(player);
            mode = oldMode == ModuleAttackCooldown.PVPMode.NEW_PVP ?
                    ModuleAttackCooldown.PVPMode.OLD_PVP : ModuleAttackCooldown.PVPMode.NEW_PVP;
        }

        final String message = config.getString("disable-attack-cooldown.message-" +
                                (mode == ModuleAttackCooldown.PVPMode.NEW_PVP ? "enabled" : "disabled"),
                        "&4ERROR: &rdisable-attack-cooldown.message strings missing")
                .replaceAll("%player%", player.getDisplayName());

        ModuleAttackCooldown.setAttackSpeed(player, mode);
        Messenger.sendNormalMessage(sender, message);
    }

    /*
    private void test(OCMMain plugin, CommandSender sender) {
        final Location location = sender instanceof Player ?
                ((Player) sender).getLocation() :
                sender.getServer().getWorlds().get(0).getSpawnLocation();

        new InGameTester(plugin).performTests(sender, location);
    }
     */

    private void wideToggle(CommandSender sender, String[] args, ModuleAttackCooldown.PVPMode mode) {
        final Set<World> worlds = args.length > 1 ?
                Arrays.asList(args).subList(1, args.length).stream().map(Bukkit::getWorld).filter(Objects::nonNull).collect(Collectors.toSet())
                : new HashSet<>(Bukkit.getWorlds());

        worlds.stream().map(World::getPlayers).forEach(players -> players.forEach(
                player -> ModuleAttackCooldown.setAttackSpeed(player, mode)));

        // Do not use method reference to get world name because with 1.18 method was moved from World to WorldInfo
        final String message = (mode == ModuleAttackCooldown.PVPMode.NEW_PVP ? "Enabled" : "Disabled") + " cooldown for worlds: " +
                worlds.stream().map(w -> w.getName()).reduce((a, b) -> a + ", " + b).orElse("none!");
        Messenger.sendNormalMessage(sender, message);
    }

    public boolean onCommand(@NotNull CommandSender sender, @NotNull Command cmd, @NotNull String label, String[] args) {
        if (args.length < 1) {
            help(plugin, sender);
        } else {
            try {
                try {
                    final Subcommand subcommand = Subcommand.valueOf(args[0].toLowerCase(Locale.ROOT));
                    if (checkPermissions(sender, subcommand, true)) {
                        switch (subcommand) {
                            case reload:
                                reload(sender);
                                break;
                            case toggle:
                                toggle(plugin, sender, args);
                                break;
                            //case test: test(plugin, sender);
                            //    break;
                            case enable:
                                wideToggle(sender, args, ModuleAttackCooldown.PVPMode.NEW_PVP);
                                break;
                            case disable:
                                wideToggle(sender, args, ModuleAttackCooldown.PVPMode.OLD_PVP);
                                break;
                            default:
                                throw new CommandNotRecognisedException();
                        }
                    }
                } catch (IllegalArgumentException e) {
                    throw new CommandNotRecognisedException();
                }
            } catch (CommandNotRecognisedException e) {
                Messenger.sendNormalMessage(sender, "Subcommand not recognised!");
            }
        }
        return true;
    }

    private static class CommandNotRecognisedException extends IllegalArgumentException {
    }

    static boolean checkPermissions(CommandSender sender, Subcommand subcommand) {
        return checkPermissions(sender, subcommand, false);
    }

    static boolean checkPermissions(CommandSender sender, Subcommand subcommand, boolean sendMessage) {
        final boolean hasPermission = sender.hasPermission("oldcombatmechanics." + subcommand);
        if (sendMessage && !hasPermission)
            Messenger.send(sender, NO_PERMISSION, "oldcombatmechanics." + subcommand);
        return hasPermission;
    }
}

File: src/main/kotlin/kernitus/plugin/OldCombatMechanics/commands/OCMCommandHandler.kt
/*
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/.
 */
package kernitus.plugin.OldCombatMechanics.commands

import kernitus.plugin.OldCombatMechanics.ModuleLoader
import kernitus.plugin.OldCombatMechanics.OCMMain
import kernitus.plugin.OldCombatMechanics.module.OCMModule
import kernitus.plugin.OldCombatMechanics.utilities.Config
import kernitus.plugin.OldCombatMechanics.utilities.Messenger
import kernitus.plugin.OldCombatMechanics.utilities.storage.PlayerStorage
import org.bukkit.Bukkit
import org.bukkit.ChatColor
import org.bukkit.command.Command
import org.bukkit.command.CommandExecutor
import org.bukkit.command.CommandSender
import org.bukkit.entity.Player
import java.util.function.Consumer

class OCMCommandHandler(private val plugin: OCMMain) : CommandExecutor {
    enum class Subcommand {
        reload, mode
    }

    private fun help(plugin: OCMMain, sender: CommandSender) {
        val description = plugin.description

        Messenger.sendNoPrefix(sender, ChatColor.DARK_GRAY.toString() + Messenger.HORIZONTAL_BAR)
        Messenger.sendNoPrefix(
            sender, "&6&lOldCombatMechanics&e by &ckernitus&e and &cRayzr522&e version &6${description.version}"
        )

        if (checkPermissions(sender, Subcommand.reload)) Messenger.sendNoPrefix(
            sender, "&eYou can use &c/ocm reload&e to reload the config file"
        )
        if (checkPermissions(sender, Subcommand.mode)) Messenger.sendNoPrefix(
            sender, Config.getConfig().getString(
                "mode-messages.message-usage"
            ) ?: "&4ERROR: &rmode-messages.message-usage string missing"
        )

        Messenger.sendNoPrefix(sender, ChatColor.DARK_GRAY.toString() + Messenger.HORIZONTAL_BAR)
    }

    private fun reload(sender: CommandSender) {
        Config.reload()
        Messenger.sendNoPrefix(sender, "&6&lOldCombatMechanics&e config file reloaded")
    }

    private fun mode(sender: CommandSender, args: Array<String>) {
        if (args.size < 2) {
            if (sender is Player) {
                val playerData = PlayerStorage.getPlayerData(sender.uniqueId)
                var modeName = playerData.getModesetForWorld(sender.world.uid)
                if (modeName.isNullOrEmpty()) modeName = "unknown"

                Messenger.send(
                    sender,
                    Config.getConfig().getString("mode-messages.mode-status")
                        ?: "&4ERROR: &rmode-messages.mode-status string missing",
                    modeName
                )
            }
            Messenger.send(
                sender,
                Config.getConfig().getString("mode-messages.message-usage")
                    ?: "&4ERROR: &rmode-messages.message-usage string missing"
            )
            return
        }

        val modesetName = args[1].lowercase()

        if (!Config.modesets.containsKey(modesetName)) {
            Messenger.send(
                sender,
                Config.getConfig().getString("mode-messages.invalid-modeset")
                    ?: "&4ERROR: &rmode-messages.invalid-modeset string missing"
            )
            return
        }

        var player: Player? = null
        if (args.size < 3) {
            if (sender is Player) {
                if (sender.hasPermission("oldcombatmechanics.mode.own")) player = sender
            } else {
                Messenger.send(
                    sender,
                    Config.getConfig().getString("mode-messages.invalid-player")
                        ?: "&4ERROR: &rmode-messages.invalid-player string missing"
                )
                return
            }
        } else if (sender.hasPermission("oldcombatmechanics.mode.others")) player = Bukkit.getPlayer(args[2])

        if (player == null) {
            Messenger.send(
                sender,
                Config.getConfig().getString("mode-messages.invalid-player")
                    ?: "&4ERROR: &rmode-messages.invalid-player string missing"
            )
            return
        }

        val worldId = player.world.uid
        val worldModesets = Config.worlds[worldId]

        // If modesets null it means not configured, so all are allowed
        if (worldModesets != null && !worldModesets.contains(modesetName)) { // Modeset not allowed in current world
            Messenger.send(
                sender,
                Config.getConfig().getString("mode-messages.invalid-modeset")
                    ?: "&4ERROR: &rmode-messages.invalid-modeset string missing"
            )
            return
        }

        val playerData = PlayerStorage.getPlayerData(player.uniqueId)
        playerData.setModesetForWorld(worldId, modesetName)
        PlayerStorage.setPlayerData(player.uniqueId, playerData)
        PlayerStorage.scheduleSave()

        Messenger.send(
            sender,
            Config.getConfig().getString("mode-messages.mode-set")
                ?: "&4ERROR: &rmode-messages.mode-set string missing",
            modesetName
        )

        // Re-apply things like attack speed and collision team
        val playerCopy: Player = player
        ModuleLoader.modules.forEach(Consumer { module: OCMModule -> module.onModesetChange(playerCopy) })
    }

    /*
    private void test(OCMMain plugin, CommandSender sender) {
        final Location location = sender instanceof Player ?
                ((Player) sender).getLocation() :
                sender.getServer().getWorlds().get(0).getSpawnLocation();

        new InGameTester(plugin).performTests(sender, location);
    }
     */
    override fun onCommand(sender: CommandSender, cmd: Command, label: String, args: Array<String>): Boolean {
        if (args.isEmpty()) {
            help(plugin, sender)
        } else {
            try {
                try {
                    val subcommand = Subcommand.valueOf(
                        args[0].lowercase()
                    )
                    if (checkPermissions(sender, subcommand, true)) {
                        when (subcommand) {
                            Subcommand.reload -> reload(sender)
                            Subcommand.mode -> mode(sender, args)
                        }
                    }
                } catch (e: IllegalArgumentException) {
                    throw CommandNotRecognisedException()
                }
            } catch (e: CommandNotRecognisedException) {
                Messenger.send(sender, "Subcommand not recognised!")
            }
        }
        return true
    }

    private class CommandNotRecognisedException : IllegalArgumentException()

    companion object {
        private const val NO_PERMISSION = "&cYou need the permission '%s' to do that!"

        @JvmOverloads
        fun checkPermissions(sender: CommandSender, subcommand: Subcommand, sendMessage: Boolean = false): Boolean {
            val hasPermission = sender.hasPermission("oldcombatmechanics.$subcommand")
            if (sendMessage && !hasPermission) Messenger.sendNoPrefix(
                sender, NO_PERMISSION, "oldcombatmechanics.$subcommand"
            )
            return hasPermission
        }
    }
}
